### PR TITLE
Optimized usage of alloca out of the loop

### LINKED ISSUE #3406: Stack overflow in OpenMP threading backend
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

### PR DESCRIPTION
**Pull request by anton-malakhov**
Fixes #3406
Partially replaces #3415

@stuartarchibald

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks @anton-malakhov, it seems like this would be a viable alternative, and if `alloca` is faster than the use of `threadprivate` then all the better. I'll give it a try with the stack busting test case and PR to your branch if it works ok.

**Comment by anton-malakhov** [comment]
@stuartarchibald I've merged your branch here but it is not marked as yours here, probably because it is in your fork

**Comment by stuartarchibald** [comment]
@anton-malakhov I'm fairly sure the unittest has a problem, apologies. I think what is happening is that the use of `@vectorize` in the module that the subprocess in the unit test imports forces instantiation of the threading layer on import. This means that delayed threading layer selection (as per the `@njit(parallel=True)` target) is skipped as the threading layer is already chosen, hence `SIGARBT` firing as there's a `fork(3)` in the process with the OMP backend already loaded.

File: numba/tests/test_parallel_backend.py [hunk 1]
```diff
@@ -499,8 +499,42 @@ def generate(cls):
...
+
+    @skip_no_omp
+    def test_omp_stack_overflow(self):
+        """
+        Tests that OMP does not overflow stack
+        """
+        runme = """if 1:
+            from numba import vectorize, threading_layer
+            import numpy as np
+            @vectorize(['f4(f4,f4,f4,f4,f4,f4,f4,f4)'], target='parallel')
```
Comment (stuartarchibald) [inline numba/tests/test_parallel_backend.py]: PEP8 for the next few lines?

File: numba/tests/test_parallel_backend.py [hunk 2]
```diff
@@ -499,8 +499,42 @@ def generate(cls):
...
+            print(threading_layer())
+        """
+        cmdline = [sys.executable, '-c', runme]
+        env = os.environ.copy()
+        env['NUMBA_THREADING_LAYER'] = "omp"
+        env['OMP_STACKSIZE'] = "100K"
+        out, err = self.run_cmd(cmdline, env=env)
+        if self._DEBUG:
+            print(out, err)
+        assert out.strip() == "omp"
```
Comment (stuartarchibald) [inline numba/tests/test_parallel_backend.py]: Perhaps check `self.assertIn("omp", out.strip())` as `STDOUT` may not be consistent?

**Comment by stuartarchibald** [comment]
Thanks for fixing up the test @anton-malakhov, this seems to all work ok now (manually tested on Linux x86_64). I made a couple of comments on the diff, once resolved this can be merge. Thanks again.

**Comment by anton-malakhov** [comment]
@stuartarchibald there is error in a different test on Windows which wasn't seen before. I believe, it should not be relevant to this PR:
```
Windows fatal exception: access violation
Current thread 0x000002a4 (most recent call first):
  File "C:\projects\numba-j46mi\numba\tests\npyufunc\test_vectorize_decor.py", line 46 in _run_and_compare
```

**Comment by stuartarchibald** [comment]
This is broken due to an LLVM mismatch. https://github.com/numba/numba/pull/3437 needs fixing+merging.

**Comment by seibert** [comment]
now that we've rolled back to LLVM 6, I think this can be merged

**Pull request merged**